app/models/test_run.rb

```diff
@@ -143,4 +143,15 @@ def unset_autotest_test_id
     return if self.in_progress?
     self.autotest_test_id = nil
   end
+
+  def autotest_test_id_uniqueness
+    return unless self.autotest_test_id
+
+    other_test_runs = TestRun.joins(role: [course: :autotest_setting])
```

Comment:
I don't think the `:autotest_setting` join is necessary? Unless you're deliberately filtering out `courses.autotest_setting_id` being `nil` here.